### PR TITLE
add Test for Tombstone Cleaning after a failure

### PR DESCRIPTION
**Pull request by krasi-georgiev**
Signed-off-by: Krasi Georgiev <kgeorgie@redhat.com>

### CODE REVIEW

File: db_test.go [hunk 1]
```diff
@@ -781,6 +783,71 @@ func TestTombstoneClean(t *testing.T) {
...
+
+		db.blocks = append(db.blocks, block)
+		expectedBlockDirs = append(expectedBlockDirs, blockDir)
+	}
+	db.compactor = &mockCompactorFailing{t: t}
+
+	// This should fail as we are using the failing compactor!
+	testutil.NotOk(t, db.CleanTombstones())
+
+	actualBlockDirs, err := blockDirs(db.dir)
```
Comment (codesome) [inline db_test.go]: Before checking if actual blocks are as expected, should we also check if after a failed compaction, the actual blocks are not as expected? Only then we can verify that the new change is working. (Something like testing if `mockCompactorFailing` is working as expected. We can do that check before returning an error in `Write` I guess)

Comment (krasi-georgiev) [reply db_test.go]: yep , I thought about this as well, but can't see how we can implement this nicely. Any  idea?

Comment (codesome) [reply db_test.go]: We can store the `expectedBlockDirs` in `mockCompactorFailing`. And just before returning the error in `Write`, we can check for not equal in the similar way how we check for equal.

I don't know how we can do that inside the actual test function. So this is a small hack around as you are already storing `t  *testing.T` object inside `mockCompactorFailing`

Comment (krasi-georgiev) [reply db_test.go]: ooh , yeah that is a good idea , Will try it now.

**Comment by krasi-georgiev** [comment]
@codesome let me know if you see anymore improvements we can add.

File: db_test.go [hunk 2]
```diff
@@ -800,46 +805,69 @@ func TestTombstoneCleanFail(t *testing.T) {
...
 		expectedBlockDirs = append(expectedBlockDirs, blockDir)
 	}
-	db.compactor = &mockCompactorFailing{t: t}
 
-	// This should fail as we are using the failing compactor!
+	// Initialize the mockCompactorFailing with a room for a single compaction iteration.
+	// mockCompactorFailing will fail on the second iteration so we can check if the cleanup works as expected.
+	db.compactor = &mockCompactorFailing{
+		t:      t,
+		blocks: db.blocks,
```
Comment (codesome) [inline db_test.go:822]: Maybe you want to send the copy of `db.blocks`, else will compactor modify the actual slice?

File: db_test.go [hunk 3]
```diff
@@ -800,46 +805,69 @@ func TestTombstoneCleanFail(t *testing.T) {
...
 		ULID:    uid,
 	}
 
-	fmt.Println(dest)
 	block := createEmptyBlock(c.t, filepath.Join(dest, meta.ULID.String()), meta)
 	c.blocks = append(c.blocks, block)
+
+	// Now check that all expected blocks are actually persisted on disk.
+	// This way we make sure that the we have some blocks that are supposed to be removed.
+	var expectedBlocks []string
```
Comment (codesome) [inline db_test.go]: My idea was to have something like `c.originalBlocks`. And inside `if len(c.blocks) >= c.max`, we can check that `c.originalBlocks` and `actualBlockDirs` don't match. But seems like this also works!

File: db_test.go (hunk 2 above)
Comment (codesome) [reply db_test.go:822]: I think it is not required as we wont modify the original slice.

**Comment by codesome** [comment]
@krasi-georgiev LGTM. Merging.

**Comment by codesome** [comment]
Sorry, closed by mistake.

File: db_test.go (hunk 3 above)
Comment (krasi-georgiev) [reply db_test.go]: hm, still don't get your idea. , feel free to add another PR with your idea if it is more logical don't mind at all.

**Pull request merged**